### PR TITLE
#166240832 Enables the user to follow and unfollow

### PR DESCRIPTION
**Pull request by truestbyheart**
### use case
* [Use case document](https://docs.google.com/document/d/1Y4c3D6VmfMij2ilYrvk-aUclsa1-zNh92i7c9Alpkdw/edit?usp=sharing)
#### What does this PR do?
 * Enables users to follow and unfollow authors.
#### Description of Task to be completed?
 * Created POST ``/api/profiles/:username/follow`` path to enable the user to follow the author.
 * Created DELETE ``/api/profiles/:username/unfollow`` path to enable the user to unfollow the author.
#### How should this be manually tested?
* Clone the repo to your local machine.
* Run ``npm install``   to install the dependencies.
* Set up the   ``.env`` as instructed by the   ``.env-example``
* Run migration to create the necessary tables by running ``npm run migrate`` on your terminal.
* Run seeds to add test data by running ``npm run seed`` on your terminal.
* Run  ``npm start ``.
* Send a  POST request to ``/api/profiles/:username/follow``  to follow the author.
* Send a  DELETE request to ``/api/profiles/:username/unfollow``  to unfollow the author.
* Send a GET request to ``/api/profiles/follower`` to get followers. 
* Send a GET request to ``/api/profiles/following`` to get following. 

#### What are the relevant pivotal tracker stories?
#[166240832](https://www.pivotaltracker.com/story/show/166240832)
#### Screenshots 
Screen Shot 2019-06-25 at 21 39 21
Screen Shot 2019-06-25 at 21 39 33
Screen Shot 2019-06-25 at 21 39 57
Screen Shot 2019-06-25 at 21 38 58
Screen Shot 2019-07-09 at 15 44 26
Screen Shot 2019-07-09 at 15 44 11
Screen Shot 2019-07-05 at 14 52 14
Screen Shot 2019-07-05 at 14 52 28

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -19,3 +19,6 @@ src/api/migrations
 
 # Seeders
 src/api/seeders
+
+#models
+src/api/models
```
Comment (alainburindi) [inline .eslintignore]: eslint!! new line at the end

File: src/api/controllers/followController.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  /**
+   * This function add the follow count to the database
+   * @param {*} req request sent by user
+   * @param {*} res response after excution of request
+   * @memberof Follow
+   * @return {object} res
+   */
+  static async follow(req, res) {
+    const { username } = req.params;
+    const follower = req.user.user.username;
```
Comment (alainburindi) [inline src/api/controllers/followController.js]: why user.user??

File: src/api/controllers/followController.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      });
+      const ZERO = 0;
+      if (checkUser.length > ZERO) {
+        await models.Follow.create({
+          follower,
+          following: username
+        });
+        res.status(status.OK).json({
+          status: 200,
+          message: `Your now following ${username}`
```
Comment (alainburindi) [inline src/api/controllers/followController.js]: you are, not your

File: src/api/controllers/followController.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        }
+      });
+      const ZERO = 0;
+      if (checkUser.length > ZERO) {
+        await models.Follow.create({
+          follower,
+          following: username
+        });
+        res.status(status.OK).json({
+          status: 200,
```
Comment (alainburindi) [inline src/api/controllers/followController.js]: Is there any reason why you used 200 on create?

File: src/api/seeders/20190621184055-user.js [hunk 5]
```diff
@@ -18,6 +18,12 @@ module.exports = {
           password: hashedPassword,
           createdAt: new Date(),
           updatedAt: new Date()
+        },{
+          username: 'daniel',
+          email: 'doe@gmail.com',
+          password: 'password23423',
```
Comment (alainburindi) [inline src/api/seeders/20190621184055-user.js]: can you hash this password?

File: .eslintignore (hunk 1 above)
Comment (MCFrank16) [inline .eslintignore]: kindly add on a new line below here

File: src/api/controllers/followController.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  static async follow(req, res) {
+    const { username } = req.params;
+    const follower = req.user.user.username;
+    try {
+      const checkUser = await models.User.findAll({
+        where: {
+          username
+        }
+      });
+      const ZERO = 0;
```
Comment (MCFrank16) [inline src/api/controllers/followController.js]: there is a file in helpers which is supposed to contain all magic numbers

File: src/api/controllers/followController.js (hunk 3 above)
Comment (MCFrank16) [inline src/api/controllers/followController.js]: is it `` your now following someone`` or ``you are following someone``

File: src/api/controllers/followController.js (hunk 4 above)
Comment (MCFrank16) [inline src/api/controllers/followController.js]: on create does the status need to be a 200 or 201??

File: src/api/controllers/followController.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      } else {
+        errorMessages(status.NOT_FOUND, res, 'user', errMessage.noUser);
+      }
+    } catch (error) {
+      errorMessages(status.BAD_REQUEST, res, 'follow', `Already following ${username}`);
+    }
+  }
+
+  /**
+ * This funbction delete the follow
```
Comment (MCFrank16) [inline src/api/controllers/followController.js]: this Function unfollow the user

File: src/api/migrations/20190625104607-create-follow.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      }
+    }).then(() => queryInterface.addConstraint('follows', ['follower', 'following'], {
+      type: 'primary key',
+      name: 'followId'
+    }))
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('follows');
+  }
+};
```
Comment (MCFrank16) [inline src/api/migrations/20190625104607-create-follow.js]: new line here

File: tests/follow.test.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import errMessage from '../src/helpers/constants/error.messages';
+
+chai.use(chaihttp);
+chai.should();
+
+describe('POST /api/profiles/:username/follow', () => {
+  it('Should return a message on successful follow', (done) => {
+    const token = generateToken({ username: 'daniel' }, env.secret);
+
+    chai
```
Comment (alainburindi) [inline tests/follow.test.js:32]: can a user follow himself????

**Review by alainburindi — CHANGES_REQUESTED** [review]

File: src/api/controllers/followController.js (hunk 2 above)
Comment (truestbyheart) [reply src/api/controllers/followController.js]: within ``req.user`` there is an object ``{user:{username: 'Allain'}}`` so in order to get it I used ``req.user.user.username``

**Review by MCFrank16 — CHANGES_REQUESTED** [review]

File: src/api/controllers/followController.js (hunk 4 above)
Comment (truestbyheart) [reply src/api/controllers/followController.js]: I have already implemented it to have 201

File: src/api/controllers/followController.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+   * @param {*} req request sent by user
+   * @param {*} res response after excution of request
+   * @memberof Follow
+   * @return {object} res
+   */
+  static async follow(req, res) {
+    const { username } = req.params;
+    const follower = req.user.user.username;
+    try {
+      const checkUser = await models.User.findAll({
```
Comment (premlogic) [inline src/api/controllers/followController.js]: why do you do this check yet you already have an authorization middleware?

File: src/api/controllers/verifyEmail.js [hunk 11]
```diff
@@ -31,7 +31,7 @@ export default class {
       } else {
         error(status.NOT_FOUND, res, 'user', errorMessages.noUser);
       }
-    } else {
+    } catch (e) {
```
Comment (premlogic) [inline src/api/controllers/verifyEmail.js]: Remember @truestbyheart 
The variable names must be specific

File: src/api/controllers/followController.js (hunk 3 above)
Comment (truestbyheart) [reply src/api/controllers/followController.js]: thanks i have changed it

File: src/api/migrations/20190625104607-create-follow.js (hunk 8 above)
Comment (truestbyheart) [reply src/api/migrations/20190625104607-create-follow.js]: already added it

File: src/docs/swagger.json [hunk 12]
```diff
@@ -14,15 +14,24 @@
   },
```
Comment (premlogic) [inline src/docs/swagger.json]: I can't see the endpoint of follow and unfollow

File: tests/follow.test.js (hunk 9 above)
Comment (truestbyheart) [reply tests/follow.test.js:32]: already solved the issue, I added a constraint  to prevent the row fro having the same user

File: tests/follow.test.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
+import chai from 'chai';
+import chaihttp from 'chai-http';
+import server from '../src/index';
+import status from '../src/helpers/constants/status.codes';
+import generateToken from '../src/helpers/tokens/generate.token';
+import env from '../src/configs/environments';
+import errorMessage from '../src/helpers/constants/error.messages';
```
Comment (premlogic) [inline tests/follow.test.js:7]: errorMessages makes more sense @truestbyheart

**Review by nkpremices — CHANGES_REQUESTED** [review]

File: tests/follow.test.js (hunk 13 above)
Comment (truestbyheart) [reply tests/follow.test.js:7]: if in my code I write  ```errorMessage.noUser``` and ```errorMessages.noUser```, for me I think the former makes more sense cause its a single error am testing for.

File: src/docs/swagger.json (hunk 12 above)
Comment (truestbyheart) [reply src/docs/swagger.json]: ``/profiles/:username/follow`` this is how the path is structured @nkpremices

File: src/api/controllers/followController.js (hunk 6 above)
Comment (truestbyheart) [reply src/api/controllers/followController.js]: that for status codes

File: src/api/controllers/followController.js (hunk 7 above)
Comment (truestbyheart) [reply src/api/controllers/followController.js]: Already updated the documentation

**Review by MCFrank16 — APPROVED** [review]

**Comment by MCFrank16** [comment]
you can do better, if you increase the coveralls

**Comment by MCFrank16** [comment]
and rebase with develop to get updated please

File: src/api/controllers/followController.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
+import models from '../models/index';
+import status from '../../helpers/constants/status.codes';
+import errorSender from '../../helpers/error.sender';
+/**
+ * @class
+ */
+export default class Follow {
```
Comment (MemunaHaruna) [inline src/api/controllers/followController.js]: @truestbyheart I think the class and model name should be nouns. Perhaps `Following` might work here.

File: src/api/controllers/followController.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+ * @param {*} res response after execution of request
+ * @memberof Follow
+ * @returns {object} res
+ */
+  static async unfollow(req, res) {
+    const { username } = req.params;
+
+    const unfollowAct = await models.Follow.destroy({
+      where: {
+        following: username
```
Comment (MemunaHaruna) [inline src/api/controllers/followController.js]: @truestbyheart  We only want to delete a specific record here, else we might delete all a user's followers.

File: src/api/controllers/followController.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+      await models.Follow.create({
+        follower,
+        following: username
+      });
+      res.status(status.CREATED).json({
+        status: 201,
+        message: `You are now following ${username}`
+      });
+    } catch (error) {
+      errorSender(status.BAD_REQUEST, res, 'follow', `Already following ${username}`);
```
Comment (MemunaHaruna) [inline src/api/controllers/followController.js]: @truestbyheart how do you know that the user is already following? I don't see a check for this.

File: src/docs/swagger.json (hunk 12 above)
Comment (premlogic) [reply src/docs/swagger.json]: cool @truestbyheart

File: src/api/controllers/verifyEmail.js [hunk 17]
```diff
@@ -8,31 +8,27 @@ import errorMessages from '../../helpers/constants/error.messages';
  */
 export default class {
```
Comment (premlogic) [inline src/api/controllers/verifyEmail.js:9]: thanks for doing this @truestbyheart 
I've been trying to do it in my own PR.

File: src/api/controllers/followController.js (hunk 16 above)
Comment (premlogic) [reply src/api/controllers/followController.js]: @MemunaHaruna we implemented this together.
the idea here is that we have as primary key the compound of both `fallower:following`.

It means there is a unique constraint on their values compounded together.

Now if the user tries to follow another twice it will violate that uniqueness and it will raise an exception that @truestbyheart is catching.

**Review by nkpremices — APPROVED** [review]

**Comment by nkpremices** [comment]
I've approved the PR @truestbyheart 
Just make sure all the comments of @MemunaHaruna are implemented

File: tests/follow.test.js (hunk 13 above)
Comment (premlogic) [reply tests/follow.test.js:7]: cool @truestbyheart

File: src/api/controllers/followController.js (hunk 16 above)
Comment (truestbyheart) [reply src/api/controllers/followController.js]: Since its compound key, it will never allow duplication but I have created a script that checks the error code for duplication and self follow error  just to prevent the user from following himself and trying to follow a user he/she is following

File: src/api/models/following.js [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+    follow.belongsTo(models.User, {
+      foreignKey: 'follower',
+    });
+    follow.belongsTo(models.User, {
+      foreignKey: 'following',
+    });
+  };
+  return follow;
+ };
+
```
Comment (alainburindi) [inline src/api/models/following.js]: new line

File: src/api/migrations/20190706134036-create-following.js [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+           name: "followId"
+         })
+         .then(() =>
+         sequelize.query(`ALTER TABLE followings ADD CONSTRAINT following_cannot_be_equal_to_follower_CHK CHECK (following <> follower)`)
+         )
+     );
+ },
+ down: (queryInterface, Sequelize) => {
+   return queryInterface.dropTable("followings");
+  }
```
Comment (alainburindi) [inline src/api/migrations/20190706134036-create-following.js:48]: new line

File: tests/verify.test.js [hunk 20]
```diff
@@ -77,7 +77,7 @@ describe('GET /verifyEmail', () => {
     chai
       .request(server)
       .get(`/api/users/verifyEmail/${verifytoken}`)
-      .end((err, res) => {
+      .end(async (err, res) => {
         res.should.have.status(status.BAD_REQUEST);
         res.body.errors.link.should.be.eq(errorMessages.emailLinkInvalid);
         done();
```
Comment (alainburindi) [inline tests/verify.test.js:83]: why did you edit touch this file?

File: src/api/controllers/followController.js [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+import status from '../../helpers/constants/status.codes';
+import errorSender from '../../helpers/error.sender';
+import errorMessage from '../../helpers/constants/error.messages';
+
+const emptyArrayLength = 0;
+
+/**
+ * @class
+ */
+export default class Following {
```
Comment (MemunaHaruna) [inline src/api/controllers/followController.js]: @truestbyheart let's have the same name for the class and file. So, the file should be `followingController`

File: src/api/controllers/followController.js [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+   * @param {object} req request sent by the user
+   * @param {object} res response after excution of user request
+   * @return {object} void
+   * @memberof Follow
+   *
+   */
+  static async getUserfollower(req, res) {
+    const { username } = req.user.user;
+
+    const follower = await sequelize.query(`
```
Comment (MemunaHaruna) [inline src/api/controllers/followController.js]: @truestbyheart Thanks for adding the methods to get a user's followers and followings. 

A few things:
- I think Sequelize works fine for these and we don't need raw SQL. Open to hearing your thoughts on it, though.
- we need to add pagination to these queries. A user could have hundreds/thousands of followers/followings and we want our queries to be as efficient as possible.

File: src/api/routes/index.js [hunk 23]
```diff
@@ -3,13 +3,15 @@ import authRouter from './authRouter';
 import resetRouter from './password.reset';
 import articleRouter from './articleRouter';
 import profileRouter from './profileRouter';
-import authenticate from '../../middlewares/authenticate';
```
Comment (MemunaHaruna) [inline src/api/routes/index.js:6]: @truestbyheart why have you removed this middleware?

File: src/api/controllers/followController.js (hunk 22 above)
Comment (truestbyheart) [reply src/api/controllers/followController.js]: The issue is sequelize only compare primary keys using the ``include`` tried everything until I resorted to using raw SQL

File: src/api/routes/index.js (hunk 23 above)
Comment (truestbyheart) [reply src/api/routes/index.js:6]: @alainburindi didn't use @MCFrank16 middleware, and I was supposed to implement this feature on his route so to avoid conflict I had to move everything to ``profileRouter``. @alainburindi said he will relay this to you.

File: tests/verify.test.js (hunk 20 above)
Comment (truestbyheart) [reply tests/verify.test.js:83]: These changes were implemented on @murediane PR so I had to implement it.

File: src/api/controllers/followController.js (hunk 22 above)
Comment (truestbyheart) [reply src/api/controllers/followController.js]: https://github.com/sequelize/sequelize/issues/2967 ready it here. And stopped here because I could find useful documentations @MemunaHaruna

**Review by truestbyheart — COMMENTED** [review]
replied  to some of the feedback

**Review by alainburindi — APPROVED** [review]

File: src/api/controllers/followController.js [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+import status from '../../helpers/constants/status.codes';
+import errorSender from '../../helpers/error.sender';
+import errorMessage from '../../helpers/constants/error.messages';
+
+const emptyArrayLength = 0;
+
+/**
+ * @class
+ */
+export default class FollowingController {
```
Comment (MemunaHaruna) [inline src/api/controllers/followController.js]: @truestbyheart I can't find the comment but I think I left a comment a few days ago about making the controller name and filename the same.

File: src/api/controllers/followController.js (hunk 22 above)
Comment (MemunaHaruna) [reply src/api/controllers/followController.js]: @truestbyheart is there a reason we are referencing users' username instead of id in the followings table?
